sklearn/tree/_criterion.pyx

```diff
@@ -445,9 +579,12 @@ cdef class ClassificationCriterion(Criterion):
                 self.weighted_n_left += w
 
         else:
+            # Reverse reset the available values alone.
+            direction = self.missing_direction
             self.reverse_reset()
+            self.move_missing(direction)
```

Comment:
Indeed we are using this to reduce the number of operations... `reverse_reset` also resets the missing direction to left. To avoid this I used a helper `reverse_reset_without_missing` which we decided not to add to avoid blow up of methods... Hence we decided to `reset`/`reverse_reset` and restore the missing to the right partition...